### ISSUE TITLE
[Debug] Dashboards

### ISSUE THREAD

**Issue by gcoeugnet** [librenms/librenms#4255]
Hi,

I'm trying to debug dashboards creation on my up to date LibreNMS. Indeed, when I add another dashboards and try to add a widget in it, I get an error "An error occurred" in the WebUI.

By looking further in the code I saw that the file "html/includes/forms/update-dashboard-config.inc.php" is called. Regarding my tests, when this function is called via POST through the WebUI to remove a widget : everything is ok, I collect a JSON string in reply of the POST. But when this fonction is called through the WebUI to add a widget, a JSON is return by the function like this one below but this JSON isn't put in response of the POST request, The ajax call then failed with "error" status instead of success. 

Here is the parameter put in the POST and its reply for removing a widget :
send datas
array(4) {
  ["type"]=>
  string(23) "update-dashboard-config"
  ["sub_type"]=>
  string(6) "remove"
  ["widget_id"]=>
  string(2) "86"
  ["dashboard_id"]=>
  string(1) "5"
}

received data :
string(60) "{
    "status": "ok",
    "message": "",
    "extra": null
}"

which is ok in this screeshot : POST reply is present !
![post reply for remove widget](https://cloud.githubusercontent.com/assets/17470139/18054735/4fb8c1e4-6e05-11e6-9b46-0c306c85e85f.png)

and here is the same POST request/reply values for an add which seems to be good but which is not put in the reply of the POST :

send datas
array(4) {
  ["type"]=>
  string(23) "update-dashboard-config"
  ["sub_type"]=>
  string(3) "add"
  ["widget_id"]=>
  string(1) "3"
  ["dashboard_id"]=>
  string(1) "5"
}

data which would be in reply of POST :
string(241) "{
    "status": "ok",
    "message": "",
    "extra": {
        "user_widget_id": 88,
        "widget_id": 88,
        "title": "Alerts",
        "widget": "alerts",
        "refresh": 60,
        "size_x": "8",
        "size_y": "4"
    }
}"

In this screenshot, the return JSON datas are not in the reply of POST, there is no reply at all.
![post reply for adding widget](https://cloud.githubusercontent.com/assets/17470139/18054736/4fde2ce0-6e05-11e6-9a83-db8282d8d49d.png)

Apache logs doesn't give more information... As the server use the same script for POST reply, it don't get why it is not working, any idea ?

**Comment by gcoeugnet** [comment]
I replace html/includes/forms/update-dashboard-config.inc.php with this :
```
<?php
header('Content-type: application/json');

$status  = 'ok';
$message = 'Test';

$response = array(
    'status'        => $status,
    'message'       => $message,
    'extra'         => $extra,
);

echo _json_encode($response);
```
The situation is the same : remove works, not add. So the problème seems to be in the calling part of the code .... I keep trying to investigate until lake of idea... :)

**Comment by gcoeugnet** [comment]
I added a "async: false" to the $.ajax POST request when adding a widget in "html/pages/front/tiles.php". Now the POST request succeed with JSON data in reply but the WebUI failed later :
![post ok with async](https://cloud.githubusercontent.com/assets/17470139/18056660/2169625a-6e0d-11e6-9350-a5609968119b.png)

I guess my problem is linked with the jquery ajax method used. Can you confirm or not this behaviour ?

**Comment by gcoeugnet** [comment]
Ok, browsing the web, I changed the code of the "add" jquery event like this :
```
        $('.place_widget').on('click',  function(event, state) {
            var widget_id = $(this).data('widget_id');
            event.preventDefault();
            $.ajax({
                type: 'POST',
                async: false,
                url: 'ajax_form.php',
                data: {type: "update-dashboard-config", sub_type: 'add', widget_id: widget_id, dashboard_id: <?php echo $vars['dashboard']['dashboard_id']; ?>},
                dataType: "json",
                success: function (data) {
                    if (data.status == 'ok') {
                        widget_dom(data.extra);
                        updatePos(gridster);
                    }
                    else {
                        $("#message").html('<div class="alert alert-info">' + data.message + '</div>');
                    }
                },
                error: function () {
                    $("#message").html('<div class="alert alert-info">An error occurred.</div>');
                }
            });
        });
```
Adding `event.preventDefault();` before the $.ajax does the trick... Can you confirm the solution before any Pull Request ?

**Comment by laf** [comment]
I can't replicate your issue on the current master branch.

If all you've changed is add `event.preventDefault();` then I don't see an issue with that and feel free to create a pull request.

**Comment by laf** [comment]
To be fair this is the same fix as for #4213 

Will PR now.

**Comment by gcoeugnet** [comment]
Yes, this is the same problem, The redirection is due to the POST without reply data. 
As I don't see any PR regarding this issue, I'll do it now.
